### ISSUE TITLE
variable autocomplete

### ISSUE THREAD

**Issue by jobace78** [BashSupport-Pro/bashsupport-pro#159]
Hi,

Does BashSupport Pro support variable autocomplete?

I post you two examples (one for BashSupport Pro and one for bundled Shell Script).

Thanks.

<img width="1440" alt="BashSupportPro" src="https://github.com/BashSupport-Pro/bashsupport-pro/assets/1135365/bc5672ff-9218-49bc-8545-077215346e85">
<img width="1440" alt="ShellScript" src="https://github.com/BashSupport-Pro/bashsupport-pro/assets/1135365/b2fab761-4dd1-4918-b0f8-4823fa281b7d">

**Comment by jansorg** [comment]
Sure it does. Type `$` and you'll get the available completions.
I recommend the manual :wink: 
https://www.bashsupport.com/manual/editor/code-completion/

In your screenshot `var...` can't be a reference to a variable, but only `$var...` or `${var...}`.

The Shell plugin is just faking it, because it's completing the available words and not based on declarations. For example, the following would complete `myName`, which is wrong:
```
echo myName is not a var
$ #<-- call completions after $
```
This page compares the features:
https://www.bashsupport.com/features/

**Comment by jobace78** [comment]
Hi,

Got it, another example:

```shell
function() {
  local var1
  local var2
  va # <- at this point there is no way to complete this variables?
}
```

Thanks

**Comment by jansorg** [comment]
If you type `$`, then the local variables will be suggested in the same way as in a global context.
Do you expect to see variable completions without a `$` prefix in the editor?

**Comment by jansorg** [comment]
In your example it's not working because `function` is not a valid name for a function... Make it `function x()` or `function_x` and it'll work

**Comment by jobace78** [comment]
Hi,

It was a quick example (not really tested, sorry).

A working one:

```shell
#!/bin/bash
_test() {
  local var1
  local var2
  var1="${1:?}"
  var2="${2:-default}"
  printf '%s\n' "var1 = ${var1}"
  printf '%s\n' "var2 = ${var2}"
}
_test blah
```

In this case when you start typing `var1` (e.g. va) at line 5, there is no completion (it's a valid variable name, it's within the scope, etc).

I don't know if this is a fake or not-so-correct behaviour from Shell Script plugin, but we're migrating to BashSource Pro plugin and this is one of the behaviours what changed to us.

This behaviour also happens in VSCode, so we're really used to.

Thanks.

**Comment by jansorg** [comment]
I see, thanks for the clarification of your use-case. 
I assumed that you wanted to refer to the variable value and not re-assign it. 
In the current version, there's completion for variable names after `export`, `declare`, etc. But currently there's no completion for assignments without such command name.

Would you expect completions like `var1=` only for local variables inside a function or also for all global variables?
I just don't want to add a lot of completions cluttering function name and command name completion.

Completing local variables for assignments seems fine and also useful to me, I could add this for the next update. In your example, items `var1=` and `var2=` would be displayed.

If you'd like to see all variables (global & local), would it be acceptable to you to press ctrl+space again or to get a setting, which is off by default?

What do you think?

AFAIK the JetBrains plugin only completes function names, everything else is faked by completing words of the current document.

**Comment by jobace78** [comment]
Hi,

Completing functions or variable names for assignment (scoped will be great, e.g. global + local scoped ones) feels very "natural" IMHO. The ctrl/cmd+space approach seems fine to me also.

BTW I _really_ enjoy the function autocomplete feature from functions defined in a sourced file (e.g. `# shellcheck source=blah.sh`).

Thanks.

**Comment by jansorg** [comment]
@jobace78 I've added this for the next update.
With you slightly modified snippet, you'll get this:
- functions first
- then local variable declarations
- then global variables declared in the current function
- then all other global variables

Variables from libraries are not shown to avoid too many useless completions. Perhaps I'll show these variables when ctrl+space is pressed a 2nd time.

I've decided to always show globals, but with above ordering applied.

![image](https://github.com/BashSupport-Pro/bashsupport-pro/assets/11473063/06a1039f-c603-4c15-a0d9-8f03abd12566)
